### PR TITLE
add function for creating hash values depending on the password

### PR DESCRIPTION
**Pull request by tineheller**

### CODE REVIEW

File: src/cryptmd5.ts [hunk 1]
```diff
@@ -66,6 +66,17 @@ export class Hash {
...
+    if('' === password) {
+        return new Hash('');
+    } else {
+        return crypt_md5(password, salt);
+    }
+}
+
+
+
 /**
```
Comment (bkircher) [inline src/cryptmd5.ts]: Hmm, not sure if this belongs to cryptmd5.ts module. This has nothing to do with neither crypt(3) nor md5.

I would rather suggest something like this at the call-site:
```ts
let plaintext: string = '';
let hash: Hash = plaintext.length > 0 ? crypt_md5(plaintext, 'salt') : new Hash('');
```

What do you think?

Comment (tineheller) [inline src/cryptmd5.ts]: I don't want it on the call-site. I think it belongs to sds because it's a janus-feature. But I could move it to sds.ts.

Comment (bkircher) [inline src/cryptmd5.ts]: Well, okay, I think you are right. It belongs to node-sds. But calling this a feature is a bit exaggerating 😄 And with call site I did not necessarily meant in a different module.

How about this then?

```ts
public static changeUser(username: string, password: cryptmd5.Hash | ''): Message;
```

This way the type checker can catch if this is an empty string or a cryptmd5.Hash.

```
git --no-pager diff
diff --git a/src/sds.ts b/src/sds.ts
index 5183a51..2ea3ea6 100644
--- a/src/sds.ts
+++ b/src/sds.ts
@@ -246,13 +246,17 @@ export class Message {
      * This message logs in the given user.
      *
      * @param {string} username The user to login. Can be with principal name (e.g., 'duckburg.mickey').
-     * @param {Hash} password The user's password hashed with crypt_md5.
+     * @param {Hash} password The user's password hashed with crypt_md5 or the empty string.
      */
-    public static changeUser(username: string, password: cryptmd5.Hash): Message {
+    public static changeUser(username: string, password: cryptmd5.Hash | ''): Message {
         let msg = new Message();
         msg.add([0, 0, 0, 0, 0, 0, 0, 0, Operation.ChangeUser]);
         msg.addString(ParameterName.User, username);
-        msg.addString(ParameterName.Password, password.value);
+        if (password instanceof cryptmd5.Hash) {
+            msg.addString(ParameterName.Password, password.value);
+        } else {
+            msg.addString(ParameterName.Password, '');
+        }
         return msg;
     }
```

Comment (tineheller) [inline src/cryptmd5.ts]: Ok, the good thing about that is, that the types are correct. No "new Hash('')" anymore. Plus it shows the user that there is something special about empty passwords. I think I would even add a comment to the documentation.
The disadvantage is, that I still have to make the decision about the type on my side before calling changeUser(). Actually I wanted to get rid of that decision... However, I think it's ok. It's probably better that way.

Comment (bkircher) [inline src/cryptmd5.ts]: > The disadvantage is, that I still have to make the decision about the type on my side before calling changeUser(). Actually I wanted to get rid of that decision...

I understand what you mean here and I'd agree agree with you. But then making such "special cases" explicit at the right place in the code justifies a little redundancy.

Done in f736baa.

**Pull request closed without merging**